overlord/snapshotstate/backend/backend.go

```diff
@@ -1081,6 +1081,11 @@ func (se *SnapshotExport) StreamTo(w io.Writer) error {
 		files = append(files, path.Base(snapshotFile.Name()))
 	}
 
+	// se.Close() closes each file in se.snapshotFiles, and Close() is the
+	// finalizer for se, so make sure to keep se alive until after every file
+	// has been read.
+	runtime.KeepAlive(se)
```

Comment:
that's a good find, but something feels a little off here. From the docs:
```
KeepAlive marks its argument as currently reachable. This ensures that the object is not freed, and its finalizer is not run, before the point in the program where KeepAlive is called. 
```
and we're marking `se` as live object, **while** we're in a function attached to the `se`'s type. I would assume that until `StreamTo` finished execution, `se` would be help alive anyway.

Reply:
This change causes the test to fail consistently:

```diff
diff --git a/overlord/snapshotstate/backend/backend.go b/overlord/snapshotstate/backend/backend.go
index 77615ca943..8c6b241d7b 100644
--- a/overlord/snapshotstate/backend/backend.go
+++ b/overlord/snapshotstate/backend/backend.go
@@ -1056,6 +1056,8 @@ func (se *SnapshotExport) StreamTo(w io.Writer) error {
 
        // write out the individual snapshots
        for _, snapshotFile := range se.snapshotFiles {
+               runtime.GC()
+               runtime.GC()
                stat, err := snapshotFile.Stat()
                if err != nil {
                        return err
```

So the problem is the GC triggering the finalizer between getting the last file and that file being fully read.

So I believe the fix is correct :)

Reply:
but then this makes it work too:
```diff
diff --git a/overlord/snapshotstate/backend/backend.go b/overlord/snapshotstate/backend/backend.go
index 77615ca943..c0b3081e72 100644
--- a/overlord/snapshotstate/backend/backend.go
+++ b/overlord/snapshotstate/backend/backend.go
@@ -1055,7 +1055,9 @@ func (se *SnapshotExport) StreamTo(w io.Writer) error {
        }
 
        // write out the individual snapshots
-       for _, snapshotFile := range se.snapshotFiles {
+       for i := range se.snapshotFiles {
+               snapshotFile := se.snapshotFiles[i]
+               runtime.GC()
                stat, err := snapshotFile.Stat()
                if err != nil {
                        return err
@@ -1080,11 +1082,10 @@ func (se *SnapshotExport) StreamTo(w io.Writer) error {
 
                files = append(files, path.Base(snapshotFile.Name()))
        }
-
        // se.Close() closes each file in se.snapshotFiles, and Close() is the
        // finalizer for se, so make sure to keep se alive until after every file
        // has been read.
-       runtime.KeepAlive(se)
+       //runtime.KeepAlive(se)
 
        // write the metadata last, then the client can use that to
        // validate the archive is complete
```

Which probably makes sense, as we're not iterating on the list but rather, the `se` object needs to be kept alive in case we are accessing list elements directly. It feels like we're only patching the symptom but not addressing the root cause realy. Merely printing out se.Size() after calling StreamTo() in the test like so:
```diff
diff --git a/overlord/snapshotstate/backend/backend_test.go b/overlord/snapshotstate/backend/backend_test.go
index adf2f578bd..862da4ca48 100644
--- a/overlord/snapshotstate/backend/backend_test.go
+++ b/overlord/snapshotstate/backend/backend_test.go
@@ -1319,6 +1319,7 @@ func (s *snapshotSuite) TestExportTwice(c *check.C) {
        // and we can stream the data
        err = se.StreamTo(buf)
        c.Check(err, check.IsNil)
+       fmt.Printf("se size %v\n", se.Size())
        c.Check(buf.Len(), check.Equals, int(expectedSize))
 
        // and again to ensure size does not change when exported again
```

will make the test succeed at least until this place. This highlights how fragile this code is and I think we should look for a slightly different way of addressing this. The daemon code keeps reference to `se` and calls `se.Close()` explicitly so it's unlikely this could happen there, but the tests don't call Close() at all and indeed, this snippet fixes the test:
```diff
diff --git a/overlord/snapshotstate/backend/backend_test.go b/overlord/snapshotstate/backend/backend_test.go
index adf2f578bd..d4cb24c6a8 100644
--- a/overlord/snapshotstate/backend/backend_test.go
+++ b/overlord/snapshotstate/backend/backend_test.go
@@ -1313,6 +1313,7 @@ func (s *snapshotSuite) TestExportTwice(c *check.C) {
        ctx := context.Background()
        se, err := backend.NewSnapshotExport(ctx, shID)
        c.Assert(err, check.IsNil)
+       defer se.Close()
        err = se.Init()
        c.Assert(err, check.IsNil)
        c.Check(se.Size(), check.Equals, expectedSize)
@@ -1330,6 +1331,7 @@ func (s *snapshotSuite) TestExportTwice(c *check.C) {
        defer restore()
        se2, err := backend.NewSnapshotExport(ctx, shID)
        c.Assert(err, check.IsNil)
+       defer se2.Close()
        err = se2.Init()
        c.Assert(err, check.IsNil)
        c.Check(se2.Size(), check.Equals, expectedSize)
```

Reply:
Thinking about this a bit more, the API of snapshot exporter is broken. The intention is to allow consumers to not call `SnapshotExporter.Close()`, either by omission or on purpose. However, in this case, StreamTo(), which should work may sometimes appear to be broken, if you're unlucky enough to have the GC make a pass. With such approach the fix is 'correct', in the sense the API is no longer broken. but the implementation remains very fragile.

Reply:
I'd tweak the comment to be more descriptive:
```
	// SnapshotExporter has a finalizer set of it, thus when the object
	// is no longer referenced, se.Close() (which closes all files) will be called automatically
	// after/during a GC pass. We don't know if the caller retains a reference to the object
	// (eg. for any outstanding calls to some of its functions), and the last explicit reference
	// in the code above was kept for the purpose of accessing the snapshot files list,
	// so we need to mark object as alive until after every file has been read.
```

Reply:
I think the explicit `runtime.KeepAlive` call is less fragile than relying on

```go
for i := range se.snapshotFiles {
    snapshotFile := se.snapshotFiles[i]
   ...
```

As Andrew mentioned below, I agree that relying on `SetFinalizer` is dangerous in general, but unfortunately Go doesn't have built-in mechanisms to prevent leaks in this way, so I'm not sure there's a way to do the same thing without using a finalizer.